### PR TITLE
[BACKLOG-26664] will:Change unification with actions

### PR DESCRIPTION
**Pull request by davidmsantos90**
@pentaho/millenniumfalcon please review

 - Simplified Actions by extending from 'Base' and moved them outside of 'pentaho/type'
 - Converted 'Changes' into 'Actions'. Both classes were very similar.

 Related PR:
 - https://github.com/pentaho/pentaho-analyzer/pull/1919

### CODE REVIEW

File: impl/client/src/test/javascript/pentaho/visual/base/View.spec.js [hunk 1]
```diff
@@ -1549,9 +1548,11 @@ define([
         };
 
         var view = new BaseView();
-        view.on(CustomSyncAction.type.id, observer);
+        var action = new CustomSyncAction();
```
Comment (dcleao) [inline impl/client/src/test/javascript/pentaho/visual/base/View.spec.js]: Just use a const var with the expected id.

File: impl/client/src/test/javascript/pentaho/visual/base/View.spec.js [hunk 2]
```diff
@@ -1575,10 +1576,11 @@ define([
         };
 
         var view = new BaseView();
+        var action = new CustomSyncAction();
 
-        view.on(CustomSyncAction.type.id, observer);
+        view.on(action.type, observer);
```
Comment (dcleao) [inline impl/client/src/test/javascript/pentaho/visual/base/View.spec.js]: idem

File: impl/client/src/test/javascript/pentaho/visual/base/View.spec.js [hunk 3]
```diff
@@ -1602,10 +1604,11 @@ define([
...
```
Comment (dcleao) [inline impl/client/src/test/javascript/pentaho/visual/base/View.spec.js]: idem

**Comment by davidmsantos90** [comment]
@dcleao pull request updated. please review

File: impl/client/src/main/javascript/web/pentaho/action/Base.js [hunk 4]
```diff
@@ -80,51 +80,96 @@ define([
...
      * @type {string}
+     * @readonly
      *
      * @abstract
      */
     get type() {
     },
 
+    /**
+     * Gets the event name of the action.
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/action/Base.js]: When `@name` is used, a doclet independent of any surrounding code is created. So, you may remove the implementation. Idem for all other abstract doclets in this file.

File: impl/client/src/main/javascript/web/pentaho/action/Base.js [hunk 5]
```diff
@@ -80,51 +80,96 @@ define([
...
      * @name clone
-     * @memberOf pentaho.action.Abstract#
+     * @memberOf pentaho.action.Base#
      *
      * @abstract
+     * @method
+     * @return {pentaho.action.Base} The action clone.
      */
     clone: function() {
+      return new this.constructor(this.toSpec());
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/action/Base.js]: Abstract or not?

File: impl/client/src/main/javascript/web/pentaho/action/Base.js [hunk 6]
```diff
@@ -80,51 +80,96 @@ define([
...
 
     /**
+     * Creates a specification that describes this action.
+     *
      * @name toSpec
-     * @memberOf pentaho.action.Abstract#
+     * @memberOf pentaho.action.Base#
      *
-     * @abstract
+     * @method
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/action/Base.js]: Abstract or not. If not, remove `@name` and `@memberOf`.

File: impl/client/src/main/javascript/web/pentaho/action/Base.js [hunk 7]
```diff
@@ -80,51 +80,96 @@ define([
...
+
+    /**
+     * Fills the given specification with this action's attributes' local values,
+     * and returns whether any attribute was actually added.
+     *
+     * @param {pentaho.action.spec.IBase} [spec] An action specification.
+     * @protected
+     * @abstract
+     */
+    _fillSpec: function(spec) {
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/action/Base.js]: Better use the same convention as other doclets and just remove the implementation, adding `@name` and `@memberOf`.

File: impl/client/src/main/javascript/web/pentaho/action/Base.js [hunk 8]
```diff
@@ -80,51 +80,96 @@ define([
...
-  }, /** @lends pentaho.action.Abstract.*/{
+  }, /** @lends pentaho.action.Base*/{
 
     /**
+     * Indicates if the action is synchronous.
+     *
+     * When unspecified, inherits the value of the ancestor action.
      *
-     * @memberOf pentaho.action.Abstract.
+     * @memberOf pentaho.action.Base
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/action/Base.js]: `@memberOf` not needed as a `@lends` is in scope.

File: impl/client/src/main/javascript/web/pentaho/action/Generic.js [hunk 9]
```diff
@@ -15,18 +15,18 @@
...
 
-  var GenericAction = AbstractAction.extend(module.id, /** @lends pentaho.action.Generic# */{
+  var GenericAction = ActionBase.extend(module.id, /** @lends pentaho.action.Generic# */{
 
     /**
      * @alias Generic
      * @memberOf pentaho.action
      * @class
-     * @extends pentaho.action.Abstract
+     * @extends pentaho.action.Base
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/action/Generic.js]: No `@classdesc` provided

File: impl/client/src/main/javascript/web/pentaho/type/events/WillChange.js [hunk 10]
```diff
@@ -37,7 +37,7 @@ define([
    * @description Creates a `WillChange` event.
    *
    * @param {pentaho.type.Complex} source - The object which is emitting the event.
-   * @param {pentaho.type.changes.Changeset} changeset -  The changeset.
+   * @param {pentaho.type.type.Changeset} changeset -  The changeset.
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/type/events/WillChange.js]: `action`

**Comment by davidmsantos90** [comment]
@dcleao pull request updated. please review
 - validate needs to be on the Base action class, in order for the execution of actions can work by default
 - clone needs to stay in the Base action class for now, it is still being called in the Execution constructor, but can be removed in the future

**Comment by davidmsantos90** [comment]
@pentaho/millenniumfalcon PR updated with all changes too Changes, Actions and Transactions

Related PRs:
 - https://github.com/pentaho/pentaho-analyzer/pull/1919
 - https://github.com/webdetails/cgg/pull/134
 - https://github.com/pentaho/pentaho-det/pull/1287
 - No changes in CDF atm

Missing unit tests

**Pull request merged**